File: Android/voiceroomcore/src/main/java/com/trtc/uikit/livekit/voiceroomcore/manager/observer/RoomEngineObserver.java
package com.trtc.uikit.livekit.voiceroomcore.manager.observer;

import com.google.gson.Gson;
import com.tencent.cloud.tuikit.engine.room.TUIRoomDefine;
import com.tencent.cloud.tuikit.engine.room.TUIRoomObserver;
import com.trtc.uikit.livekit.voiceroomcore.common.utils.Logger;
import com.trtc.uikit.livekit.voiceroomcore.manager.VoiceRoomManager;

import java.util.List;
import java.util.Map;

public class RoomEngineObserver extends TUIRoomObserver {
    private final String FILE = "LiveObserver[" + hashCode() + "]";

    protected VoiceRoomManager mVoiceRoomManager;

    public RoomEngineObserver(VoiceRoomManager manager) {
        mVoiceRoomManager = manager;
    }

    @Override
    public void onRoomDismissed(String roomId, TUIRoomDefine.RoomDismissedReason reason) {
        Logger.info(FILE, " onRoomDismissed:[roomId" + roomId + ",reason:" + reason + "]");
        mVoiceRoomManager.getRoomManager().onRoomDismissed(roomId, reason);
    }

    @Override
    public void onRoomUserCountChanged(String roomId, int userCount) {
        Logger.info(FILE, " onRoomUserCountChanged:[roomId:" + roomId + ",userCount:" + userCount + "]");
    }

    @Override
    public void onSeatListChanged(List<TUIRoomDefine.SeatInfo> seatList, List<TUIRoomDefine.SeatInfo> seatedList,
                                  List<TUIRoomDefine.SeatInfo> leftList) {
        Logger.info(FILE, " onSeatListChanged:[seatList:" + new Gson().toJson(seatList)
                + ",seatedList:" + new Gson().toJson(seatedList) + ",leftList:" + new Gson().toJson(leftList) + "]");
        mVoiceRoomManager.getSeatManager().onSeatListChanged(seatList, seatedList, leftList);
    }

    @Override
    public void onRequestReceived(TUIRoomDefine.Request request) {
        Logger.info(FILE, " onRequestReceived:[request:" + new Gson().toJson(request) + "]");
        mVoiceRoomManager.getSeatManager().onRequestReceived(request);
    }

    public void onRequestCancelled(TUIRoomDefine.Request request, TUIRoomDefine.UserInfo operateUser) {
        Logger.info(FILE, " onRequestCancelled:[request:" + request + ",operateUser:" + operateUser + "]");
        mVoiceRoomManager.getSeatManager().onRequestCancelled(request, operateUser);
    }

    @Override
    public void onRequestProcessed(TUIRoomDefine.Request request, TUIRoomDefine.UserInfo operateUser) {
        Logger.info(FILE, " onRequestProcessed:[requestId:" + request + ",operateUser:" + operateUser + "]");
        mVoiceRoomManager.getSeatManager().onRequestProcessed(request, operateUser);
    }

    @Override
    public void onKickedOffSeat(int seatIndex, TUIRoomDefine.UserInfo operateUser) {
        Logger.info(FILE, " onKickedOffSeat:[seatIndex:" + seatIndex + "]");
        mVoiceRoomManager.getSeatManager().onKickedOffSeat(seatIndex, operateUser);
    }

    @Override
    public void onUserAudioStateChanged(String userId, boolean hasAudio, TUIRoomDefine.ChangeReason reason) {
        Logger.info(FILE, " onUserAudioStateChanged:[userId:" + userId + ",hasAudio:" + hasAudio
                + ",reason:" + reason + "]");
        mVoiceRoomManager.getUserManager().onUserAudioStateChanged(userId, hasAudio, reason);
    }

    @Override
    public void onUserVoiceVolumeChanged(Map<String, Integer> volumeMap) {
        mVoiceRoomManager.getUserManager().onUserVoiceVolumeChanged(volumeMap);
    }

    @Override
    public void onRemoteUserEnterRoom(String roomId, TUIRoomDefine.UserInfo userInfo) {
        Logger.info(FILE, " onRemoteUserEnterRoom:[roomId:" + roomId + ",userId:" + userInfo.userId + "]");
    }

    @Override
    public void onRemoteUserLeaveRoom(String roomId, TUIRoomDefine.UserInfo userInfo) {
        Logger.info(FILE, " onRemoteUserLeaveRoom:[roomId:" + roomId + ",userId:" + userInfo.userId + "]");
    }

    @Override
    public void onKickedOffLine(String message) {
        Logger.info(FILE, " onKickedOffLine:[message:" + message + "]");
    }

    @Override
    public void onKickedOutOfRoom(String roomId, TUIRoomDefine.KickedOutOfRoomReason reason, String message) {
        Logger.info(FILE, " onKickedOutOfRoom:[roomId:" + roomId + ",reason:" + reason + ",message:"
                + message + "]");
        mVoiceRoomManager.getRoomManager().onKickedOutOfRoom(roomId, reason, message);
    }
}

File: Flutter/livekit/lib/voice_room/manager/observer/room_engine_observer.dart
import 'package:rtc_room_engine/api/room/tui_room_observer.dart';

import '../index.dart';

class RoomEngineObserver extends TUIRoomObserver {
  late final Context context;

  void init(Context context) {
    this.context = context;
  }

  RoomEngineObserver() {
    super.onRoomNameChanged = (roomId, roomName) {
      context.roomManager.target?.onRoomNameChanged(roomId, roomName);
    };

    super.onRoomSeatModeChanged = (roomId, seatMode) {
      context.roomManager.target?.onRoomSeatModeChanged(roomId, seatMode);
    };

    super.onRoomDismissed = (roomId, reason) {
      context.roomManager.target?.onRoomDismissed(roomId);
    };

    super.onRemoteUserEnterRoom = (roomId, userInfo) {
      context.userManager.target?.onRemoteUserEnterRoom(roomId, userInfo);
    };

    super.onRemoteUserLeaveRoom = (roomId, userInfo) {
      context.userManager.target?.onRemoteUserLeaveRoom(roomId, userInfo);
    };

    super.onRoomUserCountChanged = (roomId, userCount) {
      context.roomManager.target?.onRoomUserCountChanged(roomId, userCount);
    };

    super.onKickedOffLine = (message) {
      context.roomManager.target?.onKickedOffLine(message);
    };

    super.onKickedOutOfRoom = (roomId, reason, message) {
      context.roomManager.target?.onKickedOutOfRoom(roomId, reason, message);
    };

    super.onSeatListChanged = (seatList, seatedList, leftList) {
      context.seatManager.target
          ?.onSeatListChanged(seatList, seatedList, leftList);
    };
  }
}
